routes/index.js

```diff
@@ -110,4 +108,94 @@ router.get(`/${APP_VERSION}/datastore_search`, async function (req, res, next) {
   }
 })
 
+//TODO finish this test function to manually check downloads
+// router.get(`/test/download`, async function (req, res, next) {
+//   // res.sendFile('./test-download.html')
+//   const ppath = __dirname.split(path.sep).slice(0, -1).join(path.sep)
+//   res.sendFile(path.join(ppath + '/test/test-download.html'))
+// })
+/**
+ *
+ */
+
+router.post(`/${APP_VERSION}/download`, async function (req, res, next) {
+  console.log(' Download CALLED')
+  // get the graphql query from body
+  const query = req.body.query ? req.body.query : req.body
+  // call GraphQL
+  try {
+    // TODO check graphql syntax BEFORE sending it
+    const gqlRes = await request(`${process.env.HASURA_URL}/v1/graphql`, query)
+
+    // // capture graphql response
+    const ext = (req.params.format || req.query.format || 'json')
+      .toLowerCase()
+      .trim()
+    const colSep = (req.query.field_separator || ',').trim()
+    res.set(
+      'Content-Disposition',
+      'attachment; filename="download.' + ext + '";'
+    )
+    if (ext != 'json') {
+      // any spreadsheet supported by [js-xlsx](https://github.com/SheetJS/sheetjs)
+      let wb = XLSX.utils.book_new()
+      // TODO control the column/field order:
+      // https://stackoverflow.com/questions/56854160/sort-and-filter-columns-with-xlsx-js-after-json-to-sheet
+      // https://github.com/SheetJS/sheetjs/issues/738
+      // it needs to receive the header parameter with the desired column order
+
+      //iterate over the result sets and create a work sheet to append to the book
+      Object.keys(gqlRes).map((k) => {
+        const ws = XLSX.utils.json_to_sheet(gqlRes[k])
+        XLSX.utils.book_append_sheet(wb, ws, k)
+      })
+      if (ext === 'csv' && colSep != ',') {
+        res.set('Content-Type', 'text/csv')
+        // only send the first sheet
+        const sname = wb.SheetNames[0]
+        const ws = wb.Sheets[sname]
+        // TODO deal with record separator
+        // const recSep = (req.query.record_separator || undefined).trim() // req.params.field_separator ||
+        // const csv = XLSX.utils.sheet_to_csv(ws, { FS: colSep, RS: recSep })
+        const csv = XLSX.utils.sheet_to_csv(ws, { FS: colSep })
+        const readable = Readable.from(csv, { encoding: 'utf8' })
+        for await (const chunk of readable) {
+          if (!res.write(chunk)) {
+            // Handle backpressure
+            await once(res, 'drain')
+          }
+        }
+        res.end()
+      } else {
+        if (ext === 'csv') {
+          res.set('Content-Type', 'text/csv')
+        } else {
+          res.set(
+            'Content-Type',
+            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
```

Comment:
So, excel is basically a default one. Would it be better to explicitly say if (excel) send <something> and then if the user asked for some unsupported format (like `pdf`) return the error about the format not being supported instead?

Reply:
No, the default is JSON, that Content-Type actually works (according to the definition) for any spreadsheet format (even csv)